### PR TITLE
Give exhaustive specification of span kinds for messaging spans

### LINKED ISSUE #1112: Be more specific about span kinds for messaging operations
Based on https://github.com/open-telemetry/semantic-conventions/pull/1006#discussion_r1622108059.

Messaging semantic conventions don't give exhaustive recommendations for which span kind to use for processing and publishing operations. This can lead to inconsistencies between different instrumentations.

Semantic conventions should give exhaustive recommendations for span kinds, also for processing and publishing spans.

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes #1112, #1366

## Changes

For certain cases, messaging semantic conventions leave the span kind for `publish` and `process` spans unspecified. This is not desirable, as it can lead to inconsistent span kinds between different implementations of semantic conventions.

For all unspecified cases, this PR defines a span kind of type `CLIENT` for `publish` and `SERVER` for `process` spans.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] ~[schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.~

### CODE REVIEW

File: docs/messaging/messaging-spans.md [hunk 1]
```diff
@@ -218,20 +218,14 @@ Span kind SHOULD be set according to the following table, based on the operation
...
-the intermediary.
-
-Setting span kinds according to this table ensures that span links between
-consumers and producers always exist between a PRODUCER span on the producer
-side and a CONSUMER span on the consumer side. This allows analysis tools to
-interpret linked traces without the need for additional semantic hints.
+| `process`      | `CONSUMER` for push-based scenarios where no "Receive" span exists, otherwise `SERVER`. |
+
+Setting span kinds according to this table ensures that there always exists a
+span link between a PRODUCER span on the producer side and a CONSUMER span on
```
Comment (trask) [inline docs/messaging/messaging-spans.md:226]: 👍

**Review by trask — APPROVED** [review]

**Review by joaopgrassi — APPROVED** [review]

**Review by lmolkova — APPROVED** [review]

File: docs/messaging/messaging-spans.md [hunk 2]
```diff
@@ -216,20 +216,14 @@ Span kind SHOULD be set according to the following table, based on the operation
 | Operation type | Span kind|
 |----------------|-------------|
 | `create`       | `PRODUCER` |
-| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context, otherwise `CLIENT`. |
 | `receive`      | `CONSUMER` |
-| `process`      | `CONSUMER` for push-based scenarios where no "Receive" span exists. |
```
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:221]: Based on the discussion in the Messaging SIG 6/20 we should consider `Internal` kind

**Comment by jsuereth** [comment]
This is blocked on specification changes.

**Comment by pyohannes** [comment]
This is blocked on https://github.com/open-telemetry/opentelemetry-specification/pull/4088.

File: docs/messaging/messaging-spans.md [hunk 3]
```diff
@@ -216,20 +216,14 @@ Span kind SHOULD be set according to the following table, based on the operation
 | Operation type | Span kind|
 |----------------|-------------|
 | `create`       | `PRODUCER` |
-| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context, otherwise `CLIENT`. |
 | `receive`      | `CONSUMER` |
```
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:220]: I think it should be `CLIENT`

https://github.com/open-telemetry/semantic-conventions/issues/1366

Comment (lmolkova) [reply docs/messaging/messaging-spans.md]: updated

File: docs/messaging/messaging-spans.md (hunk 2 above)
Comment (lmolkova) [reply docs/messaging/messaging-spans.md:214]: since `recieve` is `client` now, we can always use `consumer` for the `process` span

**Review by joaopgrassi — APPROVED** [review]

**Review by trask — APPROVED** [review]

**Pull request merged**